src/v/storage/disk_log_impl.cc

```diff
@@ -557,46 +558,56 @@ model::term_id disk_log_impl::term() const {
     }
     return _segs.back()->offsets().term;
 }
+static offset_stats empty_offsets(model::offset start_offset) {
+    offset_stats ret;
+    ret.start_offset = start_offset;
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
-    if (_segs.empty()) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
-    }
-    // NOTE: we have to do this because ss::circular_buffer<> does not provide
-    // with reverse iterators, so we manually find the iterator
-    segment_set::type end;
-    segment_set::type term_start;
-    for (int i = (int)_segs.size() - 1; i >= 0; --i) {
-        auto& seg = _segs[i];
-        if (!seg->empty()) {
-            if (!end) {
-                end = seg;
-            }
-            // find term start offset
-            if (seg->offsets().term < end->offsets().term) {
-                break;
-            }
-            term_start = seg;
-        }
-    }
-    if (!end) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+    /**
+     * We have to iterate backward over segments set as we have to find last
+     * non empty segment, last segment with initialized committed offset and
+     * first offset of current term
+     */
+    const auto rend = std::make_reverse_iterator(_segs.cbegin());
+    const auto rbegin = std::find_if(
+      std::make_reverse_iterator(_segs.cend()),
+      rend,
+      [](const segment_set::type& seg) { return !seg->empty(); });
+
+    // If all segments are empty, return empty offsets.
+    if (rend == rbegin) {
+        return empty_offsets(_start_offset);
     }
+
+    auto it_last_committed = std::find_if(
+      rbegin, rend, [](const segment_set::type& seg) {
+          return seg->offsets().committed_offset >= model::offset(0);
+      });
+    // If nothing was committed, use the first segment
```

Comment:
why use the first? it seems to me that this code needs to be heavily commented.

Reply:
In case the list of segments is, e.g., `{<empty>,<empty>}`, in which case the search fails, as all commiitted offsets are negative.

Reply:
I don't see how this answers the question: why select the first segment? if all of the segments where empty like you mention, then it looks like that case would be caught above on line 584.

Reply:
sorry, I meant `[<uncommitted>]`